### PR TITLE
Differential Binarization model

### PR DESCRIPTION
**Pull request by mehtamansi29**
## Description of the Change
This PR adds Real-time Scene Text Detection with Differentiable Binarization model, Differential Binarization Model(diffbin) to KerasHub. 

## Reference
For details, see Real-time Scene Text Detection with Differentiable Binarization model ([Paper](https://arxiv.org/pdf/1911.08947))

## Colab Notebook
Usage and weight transfer [Colab](https://colab.sandbox.google.com/gist/mehtamansi29/ea4ad53f43626e2415a332034ad9e0f7/diffbin_model_final.ipynb)

## Checklist
- [x] I have added all the necessary unit tests for my change.
- [x] I have verified that my change does not break existing code and works with all backends (TensorFlow, JAX, and PyTorch).
- [x] My PR is based on the latest changes of the main branch (if unsure, rebase the code).
- [x] I have followed the Keras Hub Model contribution guidelines in making these changes.
- [x] I have followed the Keras Hub API design guidelines in making these changes.
- [x] I have signed the Contributor License Agreement.

---

### CODE REVIEW

File: keras_hub/src/models/image_text_detector_preprocessor.py [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+    def call(self, x, y=None, sample_weight=None):
+        if y is None:
+            return self.image_converter(x)
+        else:
+            # Pass bounding boxes through image converter in the dictionary
+            # with keys format standardized by core Keras.
+            output = self.image_converter(
+                {
+                    "images": x,
+                    "bounding_boxes": y,
```
Comment (sineeli) [inline keras_hub/src/models/image_text_detector_preprocessor.py:34]: I don't think we need any bounding boxes for this task.

**Review by sachinprasadhs — COMMENTED** [review]
Took high level pass and left some comments.
Also,
Make al the file names in follow the same format like other files, for db_utils and losses.py

File: keras_hub/src/models/image_text_detector_preprocessor.py [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+        '''
+        Generates postprocess function to convert probability map of 
+        model output to polygon
+        '''
+        probability_maps,threshold_maps = x["probability_maps"], x["threshold_maps"]
+        binary_maps = 1.0 / (1.0 + keras.ops.exp(-50.0 * (probability_maps - threshold_maps)))
+        outputs = keras.layers.Concatenate(axis=-1)(
+            [probability_maps, threshold_maps, binary_maps])
+        return outputs
+
```
Comment (sachinprasadhs) [inline keras_hub/src/models/image_text_detector_preprocessor.py:56]: why do we need this? we don't have `.generate` function for diff_bin right?

File: keras_hub/src/models/image_text_detector_preprocessor.py [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+
+class ImageTextDetectorPreprocessor(Preprocessor):
+    """Base class for image text detector preprocessing layers."""
+
+    def __init__(
+        self,
+        image_converter=None,
+        target_size=(640, 640),
+        shrink_ratio=0.3,
```
Comment (sachinprasadhs) [inline keras_hub/src/models/image_text_detector_preprocessor.py:18]: This will be image_size and will be handled in ImageConverter, not in this

File: keras_hub/src/models/image_text_detector_preprocessor.py (hunk 2 above)
Comment (mehtamansi29) [reply keras_hub/src/models/image_text_detector_preprocessor.py:56]: Here model output is binary_maps and we need to obtain to obtain a polygon representation. 
Earlier it was in difbin_textdetector.py file. 
But now as we are creating preprocess and postprocess  in image_text_detector_preprocessor.py, so as per the design of kerashub models.

File: keras_hub/src/models/image_text_detector_preprocessor.py (hunk 3 above)
Comment (mehtamansi29) [reply keras_hub/src/models/image_text_detector_preprocessor.py:18]: I need to use image_size for getting width and height generating mask input image.

File: keras_hub/src/models/diffbin/diffbin_backbone.py [hunk 4]
```diff
@@ -0,0 +1,220 @@
...
+            to use for the model's computations and weights.
+    """
+
+    def __init__(
+        self,
+        image_encoder,
+        fpn_channels=256,
+        head_kernel_list=[3, 2, 2],
+        image_shape=(640, 640, 3),
+        dtype=None,
```
[the remark this answers is not in the document]
Comment (sachinprasadhs) [reply keras_hub/src/models/diffbin/diffbin_backbone.py:36]: you can keep` image_shape=(None, None, 3)` for consistency

**Comment by sachinprasadhs** [comment]
Rebase the code to the latest master and resolve the gemini suggested comments

File: keras_hub/src/models/diffbin/diffbin_backbone.py [hunk 5]
```diff
@@ -0,0 +1,218 @@
...
+        fpn_channels=256,
+        head_kernel_list=None,
+        image_shape=(None, None, 3),
+        dtype=None,
+        **kwargs,
+    ):
+        if head_kernel_list is None:
+            head_kernel_list = [3, 2, 2]
+        if image_shape is None:
+            image_shape = (640, 640, 3)
```
Comment (sachinprasadhs) [inline keras_hub/src/models/diffbin/diffbin_backbone.py:42]: Why do we need this, `head_kernel_list` you can add it to default argument if it is common for all the implementations, else this can be part of the config as per the specific checkpoint.

File: keras_hub/src/models/diffbin/diffbin_backbone.py [hunk 6]
```diff
@@ -0,0 +1,220 @@
...
+        out_channels // 4,
+        kernel_size=3,
+        padding="same",
+        use_bias=False,
+        name="neck_featuremap_p2",
+        dtype=dtype,
+    )(topdown_p2)
+    featuremap_p4 = layers.UpSampling2D((4, 4), dtype=dtype)(featuremap_p4)
+    featuremap_p3 = layers.UpSampling2D((2, 2), dtype=dtype)(featuremap_p3)
+    featuremap_p2 = layers.UpSampling2D((1, 1), dtype=dtype)(featuremap_p2)
```
[the remark this answers is not in the document]
Comment (sachinprasadhs) [reply keras_hub/src/models/diffbin/diffbin_backbone.py]: Address this comment, it can be removed since it doesn't do anything

File: keras_hub/src/models/diffbin/diffbin_backbone_test.py [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+            cls=DiffBinBackbone,
+            init_kwargs=self.init_kwargs,
+            input_data=self.input_data,
+            expected_output_shape={
+                "probability_maps": (2, 640, 640, 1),
+                "threshold_maps": (2, 640, 640, 1),
+            },
+            run_mixed_precision_check=False,
+            run_quantization_check=False,
+            run_data_format_check=False,
```
Comment (sachinprasadhs) [inline keras_hub/src/models/diffbin/diffbin_backbone_test.py]: Enable these tests

File: keras_hub/src/models/diffbin/diffbin_loss_test.py [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+class TestLoss(TestCase):
+    def setUp(self):
+        super().setUp()
+        self.loss_fn = DiffBinLoss(alpha=1.0, beta=10.0)
+
+    def test_loss_zero_when_y_true_equals_y_pred(self):
+        if backend.backend() == "jax":
+            pytest.skip(
+                "JAX backend does not support this test due to NaN issues."
+            )
```
Comment (sachinprasadhs) [inline keras_hub/src/models/diffbin/diffbin_loss_test.py]: May be you should investigate and make it work on JAX as well.

File: keras_hub/src/models/diffbin/diffbin_textdetector.py [hunk 9]
```diff
@@ -0,0 +1,114 @@
...
+@keras_hub_export("keras_hub.models.DiffBinImageTextDetector")
+class DiffBinTextDetector(keras.Model):
+    """Differentiable Binarization scene text detection task.
+
+    `DiffBinImageTextDetector` tasks wrap a `keras_hub.models.DiffBinBackbone`
+    and a `keras_hub.models.Preprocessor` to create a model that can be used
+    for detecting text in natural images.
+
+    The probability map output generated by `predict()` can be translated into
+    polygon representation using `model.postprocess_to_polygons()`.
```
Comment (sachinprasadhs) [inline keras_hub/src/models/diffbin/diffbin_textdetector.py]: If this postprocess_to_polygons is not implemented, you need to update the docstring accordingly.

File: keras_hub/src/models/diffbin/diffbin_textdetector.py [hunk 10]
```diff
@@ -0,0 +1,114 @@
...
+    map_output = detector(input_data)
+    ```
+
+    `map_output` now holds a 8x224x224x3 tensor, where the last dimension
+    corresponds to the model's probability map, threshold map and binary map
+    outputs. Use `postprocess_to_polygons()` to obtain a polygon
+    representation:
+    ```python
+    detector.postprocess_to_polygons(map_output[...,0])
+    ```
```
[the remark this answers is not in the document]
Comment (sachinprasadhs) [reply keras_hub/src/models/diffbin/diffbin_textdetector.py]: same here.

File: keras_hub/src/models/diffbin/diffbin_utils.py [hunk 11]
```diff
@@ -0,0 +1,233 @@
+import keras
+
+
+def Polygon(coords):
```
Comment (sachinprasadhs) [inline keras_hub/src/models/diffbin/diffbin_utils.py:4]: Follow snake casing

File: keras_hub/src/models/image_text_detector_preprocessor.py [hunk 12]
```diff
@@ -0,0 +1,57 @@
...
+        ```
+    Returns:
+        A preprocessed image and its corresponding binary mask.
+    """
+
+    def __init__(
+        self,
+        image_converter=None,
+        image_size=(640, 640),
+        annotation_size=(640, 360),
```
Comment (sachinprasadhs) [inline keras_hub/src/models/image_text_detector_preprocessor.py]: Where are we using these two arguments?

File: keras_hub/src/models/image_text_detector_preprocessor.py [hunk 13]
```diff
@@ -0,0 +1,57 @@
+import keras
+
+from keras_hub.src.api_export import keras_hub_export
+from keras_hub.src.models.preprocessor import Preprocessor
+from keras_hub.src.utils.tensor_utils import preprocessing_function
+
+
+@keras_hub_export("keras_hub.models.ImageTextDetectorPreprocessor")
+class ImageTextDetectorPreprocessor(Preprocessor):
```
Comment (sachinprasadhs) [inline keras_hub/src/models/image_text_detector_preprocessor.py]: I don't see any difference between [ImageClassifierPreprocessor](https://github.com/keras-team/keras-hub/blob/master/keras_hub/src/models/image_classifier_preprocessor.py) and this, if it is used only for preprocessing images, may be you can just use ImageClassifierPreprocessor

**Review by sachinprasadhs — COMMENTED** [review]
Added my review comment, in the issue description add the original implementation, reference paper, training colab and end to end working demo of the trained model including post processing.

File: keras_hub/src/models/diffbin/diffbin_backbone.py (hunk 5 above)
Comment (mehtamansi29) [reply keras_hub/src/models/diffbin/diffbin_backbone.py:42]: I added head_kernel_list like this, as gemini-code-reveiw suggested(Using mutable default arguments like lists or tuples is a common pitfall in Python and can lead to unexpected behavior). 
I updated this part with config. Next commit able to see the changes.

File: keras_hub/src/models/image_text_detector_preprocessor.py (hunk 12 above)
Comment (mehtamansi29) [reply keras_hub/src/models/image_text_detector_preprocessor.py]: I'll remove annotation_size argument which I missed it. As earlier I was using diffbin_utils function with image text data preprocessor. Next commit able to see the changes

File: keras_hub/src/models/image_text_detector_preprocessor.py (hunk 13 above)
Comment (mehtamansi29) [reply keras_hub/src/models/image_text_detector_preprocessor.py]: This is because earlier I was using diffbin_utils functions(some functions like cv2,shapely created with keras.ops). But this was giving not accurate result as with cv2 and shapely result. So I used this [ImageClassifierPreprocessor](https://github.com/keras-team/keras-hub/blob/master/keras_hub/src/models/image_classifier_preprocessor.py).

File: keras_hub/src/models/diffbin/diffbin_textdetector.py (hunk 9 above)
Comment (mehtamansi29) [reply keras_hub/src/models/diffbin/diffbin_textdetector.py]: Apply the changes with relevant doc string. Next commit able to see the changes

File: keras_hub/src/models/image_text_detector_preprocessor.py (hunk 13 above)
Comment (sachinprasadhs) [reply keras_hub/src/models/image_text_detector_preprocessor.py]: Then we can subclass `ImageClassifierPreprocessor`, like other vision models

File: keras_hub/src/models/diffbin/diffbin_backbone.py [hunk 14]
```diff
@@ -0,0 +1,218 @@
...
+        # === Config ===
+        self.image_encoder = image_encoder
+        self.fpn_channels = fpn_channels
+        self.head_kernel_list = head_kernel_list
+
+    def get_config(self):
+        config = super().get_config()
+        config["fpn_channels"] = self.fpn_channels
+        config["head_kernel_list"] = self.head_kernel_list
+        config["image_encoder"] = keras.layers.serialize(self.image_encoder)
```
Comment (sachinprasadhs) [inline keras_hub/src/models/diffbin/diffbin_backbone.py]: try `keras.saving.serialize_keras_object`

**Comment by mehtamansi29** [comment]
Hi @sachinprasadhs - Can you please review this PR and let me know if any changes required ?

**Comment by mehtamansi29** [comment]
Hi @divyashreepathihalli- Can you please review this model and let me know if any changes required ?

**Review by laxmareddyp — CHANGES_REQUESTED** [review]
Hey @mehtamansi29 Sorry for delaying in response. 

I have a quick question on this model implementation licensing part..!

The official reference implementation (MhLiao/DB) has no license file. Under default copyright law, this means "all rights reserved" — the pretrained weights cannot be legally redistributed.

Before we proceed with the full code review, we need clarity on the licensing plan:

- Where are the pretrained weights coming from? The official MhLiao/DB repo (Google Drive)? MMOCR? Somewhere else?
- If from MhLiao/DB — do you have permission from the original authors to redistribute the converted weights on Kaggle under kaggle://keras/diffbin/...?
- If not — would you be open to using MMOCR's DBNet checkpoints instead? MMOCR is Apache 2.0 licensed, so weight redistribution is safe.
- Alternatively one option is that will you be able  to train new weights from scratch?

This is a blocker because redistributing unlicensed weights under the keras Kaggle org is not accepted. The architecture code in this PR is fine , but the weights need a clear open-source license trail.

Once we have clear understanding on the above part, we will take a decision and move further.

Thanks.
